### ISSUE TITLE
Generic Test Container fails to start

### ISSUE THREAD

**Issue by mitchconquer** [testcontainers/testcontainers-node#41]
Hi! I love this library, thanks for making it available!

I'm having an issue on one of my two machines but I can't figure out the root cause. I have been using the library for several months now and it was working however a few days ago, my test container wouldn't start on one of my machines. On another machine it still works well. I tried comparing the Docker versions but there doesn't seem to be an obvious difference. (I'm a relative docker n00b though.)

What's happening is that the `start()` method is throwing an error and not starting.

I'm not sure exactly what would be helpful for you to know so I'll put some relevant information below. I don't really think it's your library but I'm wondering what other dependencies I should look at to see what the issue is.

Thanks so much!

---

Here is a readout of my terminal output.

```bash
env DEBUG=testcontainers yarn app:corona test JobService.test.js
yarn run v1.15.2
$ yarn workspace corona-api-poc test JobService.test.js
$ jest JobService.test.js
Determining test suites to run...

  # TEST CONTAINERS DEBUG OUTPUT
  testcontainers INFO: Using default Docker configuration +0ms
  testcontainers INFO: Creating container for image: postgres:9.6 +21ms
  testcontainers INFO: Starting container with ID: 7b5d63d4f6cdd192a164b0a223c36cdb694a2d2b2a97fc94035978119df9bab5 +129ms
  testcontainers DEBUG: Starting container health checks +533ms

  # ERROR HAPPENS BECAUSE THE START() METHOD FAILED
{"level":"error"}
TypeError: container.getMappedPort is not a function
    at startTestDB (/Users/mitchconquer/Projects/corona/packages/corona-app/app/test/testDatabase/container.js:26:97)
    at process._tickCallback (internal/process/next_tick.js:68:7)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed.
Exit code: 1
Command: /Users/mitchconquer/.nvm/versions/node/v10.15.0/bin/node
Arguments: /usr/local/Cellar/yarn/1.15.2/libexec/lib/cli.js test JobService.test.js
Directory: /Users/mitchconquer/Projects/corona/packages/corona-app
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

I have traced this back to the `testcontainer` library's `generic-container.ts` file's `waitForContainer` method. I have run the debug logs for `testcontainer` and it seems that the container health checks are starting but not completing.

I also do not see the logs from the wait strategy's `waitUntilReady` method so I believe the error is being thrown on the `withStartupTimeout` method or the early in the `waitUntilReady` method.

**`testcontainers` Version**

From `yarn.lock`

```
testcontainers@^1.1.19:
  version "1.1.19"
  resolved "https://registry.yarnpkg.com/testcontainers/-/testcontainers-1.1.19.tgz#0062682548b99c2ea93af07c71a86264510870e9"
  integrity sha512-Aa/X92Gv0EVNRuYzxkK3AEc5hnMAgiTBhB/MpouuFTjc1Dh675Xw4KQtysy5+5S85NnQjC8RlNwshqaWC6/ikA==
  dependencies:
    byline "^5.0.0"
    debug "^4.1.1"
    dockerode "^2.5.8"
    get-port "^4.1.0"
    node-duration "^1.0.2"
    stream-to-array "^2.3.0"
    tar-fs "^2.0.0"
```

**Docker Version**

```
# docker version
Client: Docker Engine - Community
 Version:           19.03.2
 API version:       1.40
 Go version:        go1.12.8
 Git commit:        6a30dfc
 Built:             Thu Aug 29 05:26:49 2019
 OS/Arch:           darwin/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          19.03.2
  API version:      1.40 (minimum version 1.12)
  Go version:       go1.12.8
  Git commit:       6a30dfc
  Built:            Thu Aug 29 05:32:21 2019
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.2.6
  GitCommit:        894b81a4b802e4eb2a91d1ce216b8817763c29fb
 runc:
  Version:          1.0.0-rc8
  GitCommit:        425e105d5a03fabd737a126ad93d62a9eeede87f
 docker-init:
  Version:          0.18.0
  GitCommit:        fec3683
```

**`package.json`**

```json
{
  "name": "corona",
  "version": "0.1.0",
  "private": true,
  "main": "src/index.js",
  "engines": {
    "yarn": "0.27.5 || >=1.0.2",
    "node": "^10.0.0"
  },
  "dependencies": {
    "@google-cloud/debug-agent": "^3.2.0",
    "@google-cloud/logging-winston": "^0.11.1",
    "@google-cloud/pubsub": "^0.28.1",
    "@google-cloud/storage": "^2.5.0",
    "@google-cloud/trace-agent": "^3.6.1",
    "@material-ui/core": "^4.0.0-beta.0",
    "@material-ui/icons": "^3.0.2",
    "app-module-path": "^2.2.0",
    "async": "^2.6.2",
    "body-parser": "^1.19.0",
    "chalk": "^2.4.2",
    "classnames": "^2.2.6",
    "connect-busboy": "^0.0.2",
    "continuation-local-storage": "^3.2.1",
    "csv-stringify": "^5.3.0",
    "dotenv": "^8.0.0",
    "express": "^4.16.4",
    "express-winston": "^3.1.0",
    "fast-crc32c": "^1.0.4",
    "fast-csv": "^3.1.0",
    "inquirer": "^6.3.1",
    "jsrsasign": "^8.0.12",
    "multer": "^1.4.1",
    "nconf": "^0.10.0",
    "node-fetch": "^2.5.0",
    "node-sequelize-stream": "^1.0.9",
    "pg": "^7.11.0",
    "pg-hstore": "^2.3.2",
    "react": "^16.8",
    "react-dom": "^16.9",
    "react-dropzone": "^10.1.4",
    "react-router-dom": "^4.1.2",
    "react-scripts": "^3.0.0",
    "recharts": "^1.5.0",
    "request": "^2.88.0",
    "sendgrid": "^5.2.3",
    "sequelize": "^5.8.6",
    "sequelize-mock": "^0.10.2",
    "short-uuid": "^3.1.1",
    "styled-components": "^4.2.0",
    "through2": "^3.0.1",
    "unzipper": "^0.10.1",
    "winston": "^3.2.1"
  },
  "scripts": {
    "start": "node --max-old-space-size=8192 ./app/server.js",
    "dev": "nodemon ./app/server.js",
    "deploy": "gcloud app deploy app.yaml",
    "logs": "gcloud app logs tail -s default",
    "precommit": "react-scripts precommit",
    "add:mapping": "node ./app/scripts/add-mapping",
    "add:account": "node ./app/scripts/add-account",
    "logs:worker": "gcloud app logs tail -s worker",
    "test": "jest",
    "test:silent": "SILENT=true jest",
    "test:coverage": "jest --watchAll=false --coverage",
    "test:add-rev-usd": "node ./app/scripts/add-rev.js",
    "sequelize": "./node_modules/.bin/sequelize"
  },
  "proxy": "http://localhost:8080",
  "browserslist": {
    "production": [">0.2%", "not dead", "not op_mini all"],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "nodemonConfig": {
    "ignore": ["test/*", "docs/*"],
    "delay": "2500"
  },
  "devDependencies": {
    "babel-cli": "^6.26.0",
    "babel-preset-env": "^1.7.0",
    "dockerode": "^2.5.8",
    "faker": "^4.1.0",
    "http-proxy": "^1.17.0",
    "jest-serial-runner": "^1.1.0",
    "nodemon": "^1.19.0",
    "npm-run-all": "^4.1.5",
    "sequelize-cli": "^5.5.0",
    "superagent": "^5.1.0",
    "supertest": "^4.0.2",
    "testcontainers": "^1.1.19"
  }
}
```

**`container.js`**

Code that invokes `testcontainer`, failing in the `startTestDB()` method, line 31 because the container has not started. Says `getMappedPort` is not a function.

```JavaScript
'use strict';

const { getAppPath } = require('../helpers');

require('app-module-path').addPath(getAppPath());

const { GenericContainer, Wait } = require('testcontainers');
const chalk = require('chalk');
const { logger } = require('../../utils/logging');

const testSqlPort = 5432;

let container;

async function startTestDB() {
  container = await new GenericContainer('postgres', '9.6')
    .withEnv('POSTGRES_PASSWORD', 'password')
    .withExposedPorts(testSqlPort)
    .withWaitStrategy(
      Wait.forLogMessage(
        'PostgreSQL init process complete; ready for start up.'
      )
    )
    .start()
    .catch(err => logger.error(err));

  // eslint-disable-next-line no-restricted-syntax
  console.log(
    chalk.green(
      '\nTest database container created with port',
      chalk.yellow(container.getMappedPort(testSqlPort))
    )
  );

  return container;
}

function stopTestDB() {
  if (!container) {
    throw new Error('Test database container not found');
  }
  return container
    .stop()
    .then(() =>
      // eslint-disable-next-line no-restricted-syntax
      console.log(chalk.green('Test database container closed successfully'))
    )
    .catch(err =>
      // eslint-disable-next-line no-restricted-syntax
      console.log(chalk.red('Error closing test database container'), err)
    );
}

module.exports = { startTestDB, stopTestDB, container };
```

**Comment by cristianrgreco** [comment]
Hi @mitchconquer 

Really glad you like the library! 🙂 

The main point of interest to me is that it works on one machine but not the other. 

Could you confirm that both machines are using the same node/docker versions? Are the machines running the same OS? Are they both running the same node project which has package-lock or yarn.lock files (have the transitive dependencies changed)?

If it was working before, what changed?

On the machine that's not working, are you able to manually start postgres:9.6 via the Docker CLI (`docker run postgres:9.6`)? 

Sometimes the Docker daemon doesn't have enough space and needs a cleanup, you could try running `docker system prune` and try again?

**Comment by mitchconquer** [comment]
Wow thanks for the super quick follow up. I don't have my other machine here but I'll see if I can reproduce the (non) issue on my husband's computer and get some answers to the above questions. 

I also believe it is something to do with my setup but I don't see obvious differences. I'll follow up with more info.

**Comment by mitchconquer** [comment]
Amazing you led me to the right path!

When I tried to run `docker run postgres:9.6` it failed with an error 

```
docker FATAL: could not extend file "base/1/2606_fsm": No space left on device HINT: Check free disk space.
```

Some Googling led me to run this blogpost about [what to do if docker runs out of space](https://rmoff.net/post/what-to-do-when-docker-runs-out-of-space/). I ran `docker system df` and sure enough, I was over my 60GB limit with over 55GB used for local volumes.

I ran `docker volume prune` and that fixed it! 🎉

Thanks again! 😄
